src/object.d

```diff
@@ -424,6 +424,547 @@ class TypeInfo
+        static void destroy(UnqualifiedType* p)
+        {
+            p.__dtor();
+        }
+    else static if (__traits(isStaticArray, T) && is(typeof(&UnqualifiedType.init[0].__dtor)))
+        static void destroy(UnqualifiedType* p)
+        {
+            foreach_reverse (i, ref e; *p)
+                e.__dtor();
+        }
+    else
+        alias destroy = doNothing;
+
+    nothrow: pure:
+
+    private final void doNothing(UnqualifiedType*) @nogc @safe {}
+
+    // On the off chance someone calls typeid(T) == typeid(U)
+    final bool opEquals(U)(const TypeInfoImpl!U rhs) @nogc @safe
+    {
+        pragma(inline, true);
+        return is(T == U);
+    }
+
+    // Is this equal to another TypeInfo?
+    // WARNING: relies on a 1:1 mapping Type : TypeInfoImpl!Type
+    override bool opEquals(const Object rhs) @nogc @safe
+    {
+        return this is rhs
+            // Instance may be duplicated across DLLs, but has the same type.
+            || cast(const TypeInfoImpl) rhs !is null;
+    }
+
+    // On the off chance someone calls typeid(T) < typeid(U) etc.
+    final int opCmp(U)(const TypeInfoImpl!U rhs) @nogc @safe
+    {
+        pragma(inline, true);
+        return is(T == U) ? 0 : __cmp(T.stringof, U.stringof);
+    }
+
+    // Is this less than/equal/greater than another TypeInfo?
+    // WARNING: relies on a 1:1 mapping Type : TypeInfoImpl!Type
+    override int opCmp(const Object rhs) @safe
+    {
+        if (this is rhs
+                // Instance may be duplicated across DLLs, but has the same type.
+                || cast(const TypeInfoImpl) rhs !is null)
+            return 0;
+        if (auto ti = cast(const TypeInfo) rhs)
+            return __cmp(this.toString, ti.toString);
+        // TypeInfo objects are all "greater" than all other types.
+        // TODO: devise a principled way to define/justify this.
+        return 1;
+    }
+
+    override string toString() @nogc @safe { return T.stringof; }
+
+    final static size_t getHash(scope const T* p) @nogc
+    {
+        pragma(inline, true);
+        static if (is(T U == enum))
+        {
+            return __typeid!U.getHash(cast(const U*) p);
+        }
+        else static if (is(T == U[], U))
+        {
+            size_t result = 0;
+            foreach (ref e; *p)
+                result = hashOf(e, result);
+            return result;
+        }
+        else static if (__traits(isIntegral, T))
+        {
+            static if (is(T : bool))
+            {
+                return *p;
+            }
+            else static if (is(T == __vector(U), U))
+            {
+                return __typeid!U.getHash(cast(const U*) p);
+            }
+            else
+            {
+                // Knuth's multiplicative hash with the golden ratio of 2^32 or 2^64
+                static if (size_t.sizeof == 4)
+                    return *p * 2_654_435_761U;
+                else
+                    return *p * 11_400_714_819_323_198_485UL;
```

Comment:
This would fail the `TypeInfo.getHash`/`object.hashOf` equivalence tests introduced in https://github.com/dlang/druntime/pull/2243.